Flaky `GlassFishLogManagerLifeCycleTest.reconfigureWithoutResolvingLevelsWithincompleteconfiguration`

https://github.com/eclipse-ee4j/glassfish/blob/a56d44c13810baf437bba73cbac8e648751c78f4/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/GlassFishLogManagerLifeCycleTest.java#L194-L221

Have seen recently:
```
2025-01-26T08:21:39.9703043Z Both records must pass via startup queue (4 failures)
2025-01-26T08:21:39.9704392Z >--java.lang.AssertionError: record 0
2025-01-26T08:21:39.9704815Z Expected: <FINEST>
2025-01-26T08:21:39.9705101Z      but: was <INFO>
2025-01-26T08:21:39.9705451Z >--java.lang.AssertionError: record 0
2025-01-26T08:21:39.9705854Z Expected: "message0"
2025-01-26T08:21:39.9706155Z      but: was "message1"
2025-01-26T08:21:39.9706532Z >--java.lang.AssertionError: record 1
2025-01-26T08:21:39.9706915Z Expected: <INFO>
2025-01-26T08:21:39.9707189Z      but: was <FINEST>
2025-01-26T08:21:39.9707521Z >--java.lang.AssertionError: record 1
2025-01-26T08:21:39.9707903Z Expected: "message1"
2025-01-26T08:21:39.9708195Z      but: was "message0"
```

Timing issue?

Looks like records are collected in order different than expected by assertions.
From the message _Both records must pass via startup queue_ I'd not suspect the order to be crucial. I believe that implementation is rather fine here and it's test that should follow.

_Originally posted by @pzygielo in https://github.com/eclipse-ee4j/glassfish/pull/24046#discussion_r1929708843_

--- a/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/GlassFishLogger.java
+++ b/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/GlassFishLogger.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2024 Eclipse Foundation and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -507,7 +507,7 @@ public class GlassFishLogger extends Logger {
             return;
         }
         while (!isFullService(getLoggingStatus())) {
-            Thread.yield();
+            Thread.onSpinWait();
         }
         checkAndLog(record);
     }

--- a/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/GlassFishLoggerFinder.java
+++ b/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/GlassFishLoggerFinder.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022 Eclipse Foundation and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2025 Eclipse Foundation and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -19,6 +19,7 @@ package org.glassfish.main.jul;
 import java.lang.System.Logger;
 import java.lang.System.LoggerFinder;
 import java.util.ResourceBundle;
+import java.util.function.Supplier;
 
 import org.glassfish.main.jul.tracing.GlassFishLoggingTracer;
 
@@ -65,6 +66,18 @@ public class GlassFishLoggerFinder extends LoggerFinder {
         @Override
         public void log(Level level, ResourceBundle bundle, String format, Object... params) {
             logger.logrb(toJUL(level), bundle, format, params);
+        }
+
+
+        @Override
+        public void log(Level level, Supplier<String> msg, Throwable thrown) {
+            logger.log(toJUL(level), thrown, msg);
+        }
+
+
+        @Override
+        public void log(Level level, Supplier<String> msg) {
+            logger.log(toJUL(level), msg);
         }
 
 

--- a/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/handler/LogRecordBuffer.java
+++ b/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/handler/LogRecordBuffer.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2024 Eclipse Foundation and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -270,6 +270,8 @@ class LogRecordBuffer {
     }
 
     private static class CapacitySemaphore extends Semaphore {
+
+        private static final long serialVersionUID = -4575150599241117311L;
 
         public CapacitySemaphore() {
             super(0, true);

--- a/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/record/GlassFishLogRecord.java
+++ b/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/record/GlassFishLogRecord.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2024 Eclipse Foundation and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -54,7 +54,16 @@ public class GlassFishLogRecord extends LogRecord {
      * @param autodetectSource autodetect source class and method
      */
     public GlassFishLogRecord(final Level level, final String message, final boolean autodetectSource) {
-        this(new LogRecord(level, message), autodetectSource);
+        super(level, null);
+        this.threadName = Thread.currentThread().getName();
+        this.record = new LogRecord(level, message);
+        if (autodetectSource) {
+            SourceDetector.detectClassAndMethod(record);
+        } else {
+            // Prevents autodetection done passively by JUL
+            record.setSourceClassName(null);
+            record.setSourceMethodName(null);
+        }
     }
 
 
@@ -62,15 +71,11 @@ public class GlassFishLogRecord extends LogRecord {
      * Wraps the log record.
      *
      * @param record the log record
-     * @param autodetectSource autodetect source class and method
      */
-    public GlassFishLogRecord(final LogRecord record, final boolean autodetectSource) {
+    public GlassFishLogRecord(final LogRecord record) {
         super(record.getLevel(), null);
         this.threadName = Thread.currentThread().getName();
         this.record = record;
-        if (autodetectSource) {
-            SourceDetector.detectClassAndMethod(record);
-        }
     }
 
 
@@ -294,6 +299,7 @@ public class GlassFishLogRecord extends LogRecord {
         private static final Set<String> LOGGING_CLASSES = Set.of(
             "org.glassfish.main.jul.GlassFishLogger",
             "org.glassfish.main.jul.GlassFishLoggerWrapper",
+            "org.glassfish.main.jul.rotation.LogFileManager$AsyncLogger",
             // see LogDomains in GlassFish sources
             "com.sun.logging.LogDomainsLogger",
             // remaining classes are in the JDK

--- a/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/record/MessageResolver.java
+++ b/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/record/MessageResolver.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022 Eclipse Foundation and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -85,7 +85,7 @@ public final class MessageResolver {
         if (GlassFishLogRecord.class.isInstance(record)) {
             return (GlassFishLogRecord) record;
         }
-        return new GlassFishLogRecord(record, false);
+        return new GlassFishLogRecord(record);
     }
 
 

--- a/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/rotation/LogFileManager.java
+++ b/nucleus/glassfish-jul-extension/src/main/java/org/glassfish/main/jul/rotation/LogFileManager.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2024 Eclipse Foundation and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -20,19 +20,22 @@ import java.io.BufferedOutputStream;
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
-import java.lang.System.Logger;
 import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.StandardCopyOption;
 import java.time.LocalDateTime;
 import java.time.format.DateTimeFormatter;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.locks.ReentrantLock;
-import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
+import org.glassfish.main.jul.record.GlassFishLogRecord;
 import org.glassfish.main.jul.tracing.GlassFishLoggingTracer;
 
-import static java.lang.System.Logger.Level.ERROR;
-import static java.lang.System.Logger.Level.INFO;
+import static java.util.logging.Level.INFO;
+import static java.util.logging.Level.SEVERE;
 import static org.glassfish.main.jul.tracing.GlassFishLoggingTracer.trace;
 
 
@@ -49,7 +52,6 @@ import static org.glassfish.main.jul.tracing.GlassFishLoggingTracer.trace;
  * @author David Matejcek
  */
 public class LogFileManager {
-    private static final Logger LOG = System.getLogger(LogFileManager.class.getName());
 
     private static final DateTimeFormatter SUFFIX_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH-mm-ss");
 
@@ -181,10 +183,9 @@ public class LogFileManager {
      */
     public void roll() {
         lock.lock();
-        try {
+        try (AsyncLogger logger = new AsyncLogger()) {
             final boolean wasOutputEnabled = isOutputEnabled();
-            logInfoAsync(
-                () -> "Rolling the file " + this.logFile + "; output was originally enabled: " + wasOutputEnabled);
+            logger.logInfo("Rolling the file " + this.logFile + "; output was originally enabled: " + wasOutputEnabled);
             disableOutput();
             File archivedFile = null;
             try {
@@ -193,11 +194,11 @@ public class LogFileManager {
                 }
                 archivedFile = prepareAchivedLogFileTarget();
                 trace(LogFileManager.class, "Archived file: " + archivedFile);
-                moveFile(logFile, archivedFile);
+                moveFile(logFile, archivedFile, logger);
                 forceOSFilesync(logFile);
                 return;
             } catch (Exception e) {
-                logErrorAsync("Error, could not rotate log file " + logFile, e);
+                logger.logError("Error, could not rotate log file " + logFile, e);
             } finally {
                 if (wasOutputEnabled) {
                     enableOutput();
@@ -322,8 +323,8 @@ public class LogFileManager {
     }
 
 
-    private void moveFile(final File logFileToArchive, final File target) throws IOException {
-        logInfoAsync(() -> "Archiving file " + logFileToArchive + " to " + target);
+    private void moveFile(final File logFileToArchive, final File target, final AsyncLogger logger) throws IOException {
+        logger.logInfo("Archiving file " + logFileToArchive + " to " + target);
         try {
             Files.move(logFileToArchive.toPath(), target.toPath(), StandardCopyOption.ATOMIC_MOVE);
         } catch (UnsupportedOperationException | IOException e) {
@@ -331,7 +332,7 @@ public class LogFileManager {
             // Windows because of multiple file handles opened. We go through Plan B to
             // copy bytes explicitly to a renamed file.
             // Can happen on some windows file systems - then we try non-atomic version at least.
-            logErrorAsync(String.format(
+            logger.logError(String.format(
                 "File %s could not be renamed to %s atomically, now trying to move it without this request.",
                 logFileToArchive, target), e);
             Files.move(logFileToArchive.toPath(), target.toPath());
@@ -345,25 +346,68 @@ public class LogFileManager {
      * any records until it finishes rolling the file.
      * <p>
      * The count of messages is limited, so we can do this.
-     */
-    private void logInfoAsync(final Supplier<String> message) {
-        trace(getClass(), message);
-        new Thread(() -> LOG.log(INFO, message), "LogFileManager-Async-Info-Logger").start();
-    }
-
-
-    /**
-     * This logs in a separate thread to avoid deadlocks. The separate thread can be blocked when
-     * the LogRecordBuffer is full while the LogFileManager is still locked and doesn't process
-     * any records until it finishes rolling the file.
-     * <p>
-     * The count of messages is limited, so we can do this.
      * <p>
      * However it is not suitable for all errors - if we cannot write to the file, this would just create
      * another record which could not be written.
      */
-    private void logErrorAsync(final String message, final Exception exception) {
-        GlassFishLoggingTracer.error(getClass(), message, exception);
-        new Thread(() -> LOG.log(ERROR, message, exception), "LogFileManager-Async-Error-Logger").start();
+    private static class AsyncLogger extends Thread implements AutoCloseable {
+
+        private final AtomicBoolean stop;
+        private final ConcurrentLinkedQueue<AsyncLogRecord> queue;
+        private final Logger logger;
+
+        private AsyncLogger() {
+            super("LogFileManagerAsyncLogger");
+            setDaemon(true);
+            this.queue = new ConcurrentLinkedQueue<>();
+            this.stop = new AtomicBoolean();
+            this.logger = Logger.getLogger(LogFileManager.class.getName(), null);
+            start();
+        }
+
+        void logInfo(final String message) {
+            trace(getClass(), message);
+            queue.add(new AsyncLogRecord(INFO, message, null));
+        }
+
+        void logError(final String message, final Exception exception) {
+            GlassFishLoggingTracer.error(getClass(), message, exception);
+            queue.add(new AsyncLogRecord(SEVERE, message, exception));
+        }
+
+        @Override
+        public void close() {
+            this.stop.set(true);
+        }
+
+        @Override
+        public void run() {
+            while(!stop.get()) {
+                drainQueue();
+                Thread.onSpinWait();
+            }
+            drainQueue();
+        }
+
+        private void drainQueue() {
+            while (true) {
+                AsyncLogRecord record = queue.poll();
+                if (record == null) {
+                    break;
+                }
+                logger.log(record);
+            }
+        }
+    }
+
+
+    private static class AsyncLogRecord extends GlassFishLogRecord {
+
+        private static final long serialVersionUID = -8159574547676058852L;
+
+        AsyncLogRecord(Level level, String message, Throwable error) {
+            super(level, message, true);
+            setThrown(error);
+        }
     }
 }

--- a/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/cfg/LoggingPropertiesTest.java
+++ b/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/cfg/LoggingPropertiesTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2023 Eclipse Foundation and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -38,7 +38,7 @@ import static org.junit.jupiter.api.Assertions.assertNotNull;
  */
 public class LoggingPropertiesTest {
 
-    public static final int PROPERTY_COUNT = 12;
+    public static final int PROPERTY_COUNT = 13;
 
     @Test
     void conversions() throws Exception {

--- a/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/handler/GlassFishLogHandlerTest.java
+++ b/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/handler/GlassFishLogHandlerTest.java
@@ -17,10 +17,13 @@
 package org.glassfish.main.jul.handler;
 
 import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.util.logging.Level;
 import java.util.logging.LogManager;
 import java.util.logging.Logger;
+import java.util.stream.Stream;
 
 import org.glassfish.main.jul.env.LoggingSystemEnvironment;
 import org.glassfish.main.jul.formatter.OneLineFormatter;
@@ -44,6 +47,7 @@ import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertNotSame;
 import static org.junit.jupiter.api.Assertions.assertSame;
 import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
 
 /**
  * @author David Matejcek
@@ -51,8 +55,8 @@ import static org.junit.jupiter.api.Assertions.assertTrue;
 @TestMethodOrder(OrderAnnotation.class)
 public class GlassFishLogHandlerTest {
 
-    private static final long MILLIS_FOR_PUMP = 20L;
     private static GlassFishLogHandler handler;
+    private static File logFile;
 
     @BeforeAll
     public static void initEnv() throws Exception {
@@ -60,7 +64,7 @@ public class GlassFishLogHandlerTest {
         LogManager.getLogManager().reset();
         LoggingSystemEnvironment.initialize();
         final GlassFishLogHandlerConfiguration cfg = new GlassFishLogHandlerConfiguration();
-        final File logFile = File.createTempFile(GlassFishLogHandlerTest.class.getCanonicalName(), ".log");
+        logFile = File.createTempFile(GlassFishLogHandlerTest.class.getCanonicalName(), ".log");
         logFile.delete();
         logFile.deleteOnExit();
         cfg.setLogFile(logFile);
@@ -107,11 +111,12 @@ public class GlassFishLogHandlerTest {
 
         System.out.println("Tommy, can you hear me?");
         // output stream is pumped in parallel to the error stream, order is not guaranteed between streams
-        Thread.sleep(MILLIS_FOR_PUMP);
+        waitForSize(logFile, 1);
         Logger.getLogger("some.usual.logger").info("Some info message");
+        waitForSize(logFile, 2);
         System.err.println("Can you feel me near you?");
         System.err.println("Příliš žluťoučký kůň úpěl ďábelské ódy");
-        Thread.sleep(MILLIS_FOR_PUMP);
+        waitForSize(logFile, 4);
         assertAll(
             () -> assertTrue(handler.isReady(), "handler.ready"),
             () -> assertTrue(cfg.getLogFile().exists(), "file exists"),
@@ -135,7 +140,7 @@ public class GlassFishLogHandlerTest {
         assertTrue(handler.isReady(), "handler.ready");
         handler.publish(new GlassFishLogRecord(Level.SEVERE, "File one, record one", false));
         // pump is now to play
-        Thread.sleep(MILLIS_FOR_PUMP);
+        waitForSize(logFile, 5);
         assertAll(
             () -> assertTrue(handler.isReady(), "handler.ready"),
             () -> assertTrue(handler.getConfiguration().getLogFile().exists(), "file one exists"),
@@ -150,8 +155,8 @@ public class GlassFishLogHandlerTest {
             )
         );
         handler.roll();
-        // There will be an asynchronous thread.
-        Thread.sleep(10L);
+        // There will be an asynchronous thread archiving the file.
+        waitForSize(logFile, 2);
         assertAll(
             () -> assertTrue(handler.isReady(), "handler.ready"),
             () -> assertTrue(handler.getConfiguration().getLogFile().exists(), "file exists"),
@@ -160,11 +165,12 @@ public class GlassFishLogHandlerTest {
                     stringContainsInOrder("INFO", "org.glassfish.main.jul.rotation.LogFileManager.roll",
                         "Rolling the file ", ".log", "output was originally enabled: true"),
                     stringContainsInOrder("INFO", "org.glassfish.main.jul.rotation.LogFileManager.moveFile",
-                        "Archiving file ", ".log", " to ", ".log_")                )
+                        "Archiving file ", ".log", " to ", ".log_")
+                    )
             )
         );
         handler.publish(new GlassFishLogRecord(Level.SEVERE, "File two, line two", false));
-        Thread.sleep(MILLIS_FOR_PUMP);
+        waitForSize(logFile, 3);
         assertAll(
             () -> assertTrue(handler.isReady(), "handler.ready"),
             () -> assertTrue(handler.getConfiguration().getLogFile().exists(), "file exists"),
@@ -200,5 +206,22 @@ public class GlassFishLogHandlerTest {
         final GlassFishLogHandlerConfiguration cfg = GlassFishLogHandler
             .createGlassFishLogHandlerConfiguration(GlassFishLogHandler.class);
         assertNotNull(cfg, "cfg");
+    }
+
+
+    private static void waitForSize(File file, long minLineCount) throws IOException {
+        long start = System.currentTimeMillis();
+        do {
+            Thread.onSpinWait();
+            final long lineCount;
+            try (Stream<String> stream = Files.lines(file.toPath(), StandardCharsets.UTF_8)) {
+                lineCount = stream.count();
+            }
+            if (lineCount >= minLineCount) {
+                return;
+            }
+        } while (System.currentTimeMillis() - start < 2000L);
+        fail("Timeout waiting until the file " + file + " grows to " + minLineCount + " lines. File content: \n"
+            + Files.readString(file.toPath(), handler.getConfiguration().getEncoding()));
     }
 }

--- a/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/handler/GlassFishLogHandlerTest.java
+++ b/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/handler/GlassFishLogHandlerTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2024 Contributors to the Eclipse Foundation
+ * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -157,9 +157,10 @@ public class GlassFishLogHandlerTest {
             () -> assertTrue(handler.getConfiguration().getLogFile().exists(), "file exists"),
             () -> assertThat("file content", Files.readAllLines(handler.getConfiguration().getLogFile().toPath()),
                 contains(
-                    stringContainsInOrder("INFO", "Rolling the file ", ".log", "output was originally enabled: true"),
-                    stringContainsInOrder("INFO", "Archiving file ", ".log", " to ", ".log_")
-                )
+                    stringContainsInOrder("INFO", "org.glassfish.main.jul.rotation.LogFileManager.roll",
+                        "Rolling the file ", ".log", "output was originally enabled: true"),
+                    stringContainsInOrder("INFO", "org.glassfish.main.jul.rotation.LogFileManager.moveFile",
+                        "Archiving file ", ".log", " to ", ".log_")                )
             )
         );
         handler.publish(new GlassFishLogRecord(Level.SEVERE, "File two, line two", false));

--- a/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/handler/LoggingPrintStreamTest.java
+++ b/nucleus/glassfish-jul-extension/src/test/java/org/glassfish/main/jul/handler/LoggingPrintStreamTest.java
@@ -98,7 +98,7 @@ public class LoggingPrintStreamTest {
             () -> assertEquals(1L, record.getThreadID()),
             () -> assertEquals(Thread.currentThread().getName(), record.getThreadName()),
             () -> assertSame(exception, record.getThrown()),
-            () -> assertNull(record.getSourceMethodName())
+            () -> assertEquals("testPrintStacktrace", record.getSourceMethodName())
         );
     }
 

--- a/nucleus/glassfish-jul-extension/src/test/resources/logging.properties
+++ b/nucleus/glassfish-jul-extension/src/test/resources/logging.properties
@@ -5,6 +5,7 @@ java.util.logging.ConsoleHandler.filter=org.glassfish.main.jul.test.TestFilter
 
 org.glassfish.main.jul.handler.GlassFishLogHandler.formatter=org.glassfish.main.jul.formatter.OneLineFormatter
 org.glassfish.main.jul.handler.GlassFishLogHandler.formatter.size.class=1
+org.glassfish.main.jul.handler.GlassFishLogHandler.formatter.printSource=true
 
 .level=FINEST
 org.glassfish.main.jul.GlassFishLogManagerLifeCycleTest.level=FINE

